### PR TITLE
Add more dir api for Deno

### LINKED ISSUE #3488: want: more dir api for system
Deno relies on https://crates.io/crates/dirs.

It provides a directory about the system, it is very useful, `Deno.home()` also depends on it

But currently Deno does not provide APIs for other directories. For example `cache directory`/`configuration directory`

Can we add to Deno's API?

```ts
console.log(Deno.homeDir());
consnole.log (Deno.configDir());
consnole.log (Deno.cacheDir());
// ...
```

- [x] homeDir()
- [ ] cacheDir()
- [ ] configDir()
- [ ] downloadDir()
...

Or it should be added to the `std`?

If this is necessary, let me know

**BTW. `Golang` also provides such functions in the standard library**

```go
// UserCacheDir returns the default root directory to use for user-specific
// cached data. Users should create their own application-specific subdirectory
// within this one and use that.
//
// On Unix systems, it returns $XDG_CACHE_HOME as specified by
// https://standards.freedesktop.org/basedir-spec/basedir-spec-latest.html if
// non-empty, else $HOME/.cache.
// On Darwin, it returns $HOME/Library/Caches.
// On Windows, it returns %LocalAppData%.
// On Plan 9, it returns $home/lib/cache.
//
// If the location cannot be determined (for example, $HOME is not defined),
// then it will return an error.
func UserCacheDir() (string, error) {
	var dir string

	switch runtime.GOOS {
	case "windows":
		dir = Getenv("LocalAppData")
		if dir == "" {
			return "", errors.New("%LocalAppData% is not defined")
		}

	case "darwin":
		dir = Getenv("HOME")
		if dir == "" {
			return "", errors.New("$HOME is not defined")
		}
		dir += "/Library/Caches"

	case "plan9":
		dir = Getenv("home")
		if dir == "" {
			return "", errors.New("$home is not defined")
		}
		dir += "/lib/cache"

	default: // Unix
		dir = Getenv("XDG_CACHE_HOME")
		if dir == "" {
			dir = Getenv("HOME")
			if dir == "" {
				return "", errors.New("neither $XDG_CACHE_HOME nor $HOME are defined")
			}
			dir += "/.cache"
		}
	}

	return dir, nil
}

// UserConfigDir returns the default root directory to use for user-specific
// configuration data. Users should create their own application-specific
// subdirectory within this one and use that.
//
// On Unix systems, it returns $XDG_CONFIG_HOME as specified by
// https://specifications.freedesktop.org/basedir-spec/basedir-spec-latest.html if
// non-empty, else $HOME/.config.
// On Darwin, it returns $HOME/Library/Application Support.
// On Windows, it returns %AppData%.
// On Plan 9, it returns $home/lib.
//
// If the location cannot be determined (for example, $HOME is not defined),
// then it will return an error.
func UserConfigDir() (string, error) {
	var dir string

	switch runtime.GOOS {
	case "windows":
		dir = Getenv("AppData")
		if dir == "" {
			return "", errors.New("%AppData% is not defined")
		}

	case "darwin":
		dir = Getenv("HOME")
		if dir == "" {
			return "", errors.New("$HOME is not defined")
		}
		dir += "/Library/Application Support"

	case "plan9":
		dir = Getenv("home")
		if dir == "" {
			return "", errors.New("$home is not defined")
		}
		dir += "/lib"

	default: // Unix
		dir = Getenv("XDG_CONFIG_HOME")
		if dir == "" {
			dir = Getenv("HOME")
			if dir == "" {
				return "", errors.New("neither $XDG_CONFIG_HOME nor $HOME are defined")
			}
			dir += "/.config"
		}
	}

	return dir, nil
}

// UserHomeDir returns the current user's home directory.
//
// On Unix, including macOS, it returns the $HOME environment variable.
// On Windows, it returns %USERPROFILE%.
// On Plan 9, it returns the $home environment variable.
func UserHomeDir() (string, error) {
	env, enverr := "HOME", "$HOME"
	switch runtime.GOOS {
	case "windows":
		env, enverr = "USERPROFILE", "%userprofile%"
	case "plan9":
		env, enverr = "home", "$home"
	}
	if v := Getenv(env); v != "" {
		return v, nil
	}
	// On some geese the home directory is not always defined.
	switch runtime.GOOS {
	case "android":
		return "/sdcard", nil
	case "darwin":
		if runtime.GOARCH == "arm" || runtime.GOARCH == "arm64" {
			return "/", nil
		}
	}
	return "", errors.New(enverr + " is not defined")
}
```

### PR DESCRIPTION
**Pull request by axetroy**
close #3488 

depend on https://crates.io/crates/dirs

- [x] Deno.cacheDir()
- [x] Deno.configDir()
- [x] Deno.dataDir()
- [x] Deno.dataLocalDir()
- [x] Deno.audioDir()
- [x] Deno.desktopDir()
- [x] Deno.documentDir()
- [x] Deno.downloadDir()
- [x] Deno.fontDir()
- [x] Deno.pictureDir()
- [x] Deno.publicDir()
- [x] Deno.templateDir()
- [x] Deno.videoDir()

If the directory does not exist, an exception is thrown

E.g. `Deno.downloadDir()`

Not every Linux has a download directory

TODO:

- [x] test case

### CODE REVIEW

File: cli/js/os.ts [hunk 1]
```diff
@@ -140,6 +140,162 @@ export function homeDir(): string {
   return path;
 }
 
+/**
+ * Returns the current user's cache directory.
+ * Requires the `--allow-env` flag.
+ */
+export function cacheDir(): string {
+  const path = sendSync(dispatch.OP_CACHE_DIR);
```
Comment (afinch7) [inline cli/js/os.ts]: Maybe combine all these into a single op?
``` TypeScript
const path = sendSync(dispatch.OP_GET_DIR, { type: "cache" })
```

Comment (axetroy) [reply cli/js/os.ts]: done! but `type` is a rust keyword. so I use `{ name: "cache" }` instead

File: cli/ops/os.rs [hunk 2]
```diff
@@ -57,17 +57,93 @@ fn op_start(
...
+    "template" => dirs::template_dir()
+      .unwrap_or_default()
+      .into_os_string()
+      .into_string()
+      .unwrap_or_default(),
+    "video" => dirs::video_dir()
+      .unwrap_or_default()
+      .into_os_string()
+      .into_string()
+      .unwrap_or_default(),
```
Comment (ry) [inline cli/ops/os.rs]: It seems you could reduce code here by lifting the unwrap_or_default().into_os_string()... into the next statement.

File: cli/ops/os.rs [hunk 3]
```diff
@@ -57,18 +58,42 @@ fn op_start(
...
+    "audio" => dirs::audio_dir(),
+    "desktop" => dirs::desktop_dir(),
+    "document" => dirs::document_dir(),
+    "download" => dirs::download_dir(),
+    "font" => dirs::font_dir(),
+    "picture" => dirs::picture_dir(),
+    "public" => dirs::public_dir(),
+    "template" => dirs::template_dir(),
+    "video" => dirs::video_dir(),
+    _ => None,
```
Comment (hayd) [inline cli/ops/os.rs]: Is this case tested?

Comment (axetroy) [reply cli/ops/os.rs]: No test cases needed here.

I change this panic when not match.

So if not match case, an exception will be thrown and the CI won't pass

**Comment by ry** [comment]
I think it’s better not to panic - because users can still trigger that by calling the op directly. Better to return Err(...) 

I’m not at my computer to give a good example but you should be able to look through the other ops and find an example of returning an error.

File: cli/js/os.ts [hunk 4]
```diff
@@ -133,13 +133,247 @@ export function start(preserveDenoNamespace = true, source?: string): Start {
...
+ * | ------- | ----------------------------------- | ---------------------------- |
+ * | Linux   | `$XDG_CACHE_HOME` or `$HOME`/.cache | /home/alice/.cache           |
+ * | macOS   | `$HOME`/Library/Caches              | /Users/Alice/Library/Caches  |
+ * | Windows | `{FOLDERID_LocalAppData}`           | C:\Users\Alice\AppData\Local |
+ */
+export function cacheDir(): string {
+  const path = sendSync(dispatch.OP_GET_DIR, { name: "cache" });
+  if (!path) {
+    throw new Error("Could not get user cache directory.");
+  }
```
Comment (ry) [inline cli/js/os.ts]: All of these places where you throw new Error can be removed if you return the error from the op.

File: cli/ops/os.rs (hunk 3 above)
Comment (hayd) [reply cli/ops/os.rs]: I think it should be tested to confirm it returns an error (not panics)

File: cli/js/os.ts (hunk 4 above)
Comment (axetroy) [reply cli/js/os.ts]: done

File: cli/ops/os.rs (hunk 3 above)
Comment (axetroy) [reply cli/ops/os.rs]: Now it returns an error

But I'm not sure where this test case should be written

Maybe `cli/js/os_test.ts` or` cli/js/dispatch_json.ts`?

**Comment by axetroy** [comment]
@ry 
I don't think OP should be open for users
This should only be allowed to be called inside Deno

**Pull request merged**